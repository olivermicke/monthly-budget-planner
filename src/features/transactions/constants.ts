import { Currency } from './types';

export const CURRENCIES: { [key in Currency['name']]: Currency } = {
  'Australian Dollar': {
    name: 'Australian Dollar',
    symbol: 'AU$',
  },
  'British Pound': {
    name: 'British Pound',
    symbol: '£',
  },
  'Canadian Dollar': {
    name: 'Canadian Dollar',
    symbol: 'CA$',
  },
  Euro: {
    name: 'Euro',
    symbol: '€',
  },
  'Japanese Yen': {
    name: 'Japanese Yen',
    symbol: '¥',
  },
  'Swiss Franc': { name: 'Swiss Franc', symbol: 'CHF' },
  'US Dollar': {
    name: 'US Dollar',
    symbol: '$',
  },
};
