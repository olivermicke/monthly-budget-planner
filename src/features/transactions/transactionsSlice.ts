import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { RootState } from '../../store';

import { Transaction } from './types';

export const slice = createSlice({
  name: 'transactions',
  initialState: {} as { [key: string]: Transaction },
  reducers: {
    add: (state, action: PayloadAction<Transaction>) => {
      const {
        amount,
        dueDay,
        isDistributedDaily,
        name,
        ...rest
      } = action.payload;

      // TODO: Round amount

      state[name] = {
        amount,
        dueDay: isDistributedDaily ? null : dueDay,
        isDistributedDaily,
        name,
        ...rest
      };
    }
    // TODO: add missing actions
    // edit:
    // remove:
  }
});

export const selectTransactions = (state: RootState) => state.transactions;

export const { add } = slice.actions;

export const reducer = slice.reducer;
