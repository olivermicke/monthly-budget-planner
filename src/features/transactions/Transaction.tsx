import React from 'react';
import { useSelector } from 'react-redux';
import { Stat, StatArrow, StatHelpText, StatLabel, StatNumber, Text } from '@chakra-ui/core';

import { selectCurrency } from './transactionsSlice';

import { Transaction as TransactionType } from './types';

export const Transaction = ({ amount, isDistributedDaily, name, type }: TransactionType) => {
  const currency = useSelector(selectCurrency);

  return (
    <Stat>
      <StatLabel>
        {name}
        {isDistributedDaily && (
          <Text as='span' color='gray.400'>
            {' '}
            [daily]
          </Text>
        )}
      </StatLabel>
      <StatHelpText alignItems='center' display='flex'>
        <StatArrow type={type === 'incoming' ? 'increase' : 'decrease'} />
        <StatNumber as='span'>
          {isDistributedDaily ? (amount / 31).toFixed(2) : amount}
          {currency.symbol}
        </StatNumber>
      </StatHelpText>
    </Stat>
  );
};
