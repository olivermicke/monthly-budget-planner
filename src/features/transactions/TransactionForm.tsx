import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { range } from 'ramda';
import {
  Box,
  Button,
  Checkbox,
  FormControl,
  FormHelperText,
  FormLabel,
  Heading,
  Input,
  NumberInput,
  NumberInputField,
  Select,
} from '@chakra-ui/core';

import { addTransaction, selectCurrency } from './transactionsSlice';

import { Transaction } from './types';

const DEFAULT_NAME = '';
const DEFAULT_AMOUNT = 31;
const DEFAULT_DUE_DAY_NUMBER = 1;
const DEFAULT_IS_DISTRIBUTED_DAILY = false;
const DEFAULT_TYPE: Transaction['type'] = 'outgoing';

const formItemProps = {
  marginTop: ['1rem', '1rem', '2rem'],
};

export const TransactionForm = () => {
  const currency = useSelector(selectCurrency);
  const dispatch = useDispatch();

  const [name, setName] = useState(DEFAULT_NAME);
  const [amount, setAmount] = useState(DEFAULT_AMOUNT);
  const [dueDayNumber, setDueDayNumber] = useState(DEFAULT_DUE_DAY_NUMBER);
  const [isDistributedDaily, setIsDistributedDaily] = useState(DEFAULT_IS_DISTRIBUTED_DAILY);
  const [type, setType] = useState<Transaction['type']>(DEFAULT_TYPE);

  return (
    <Box marginBottom={['2rem', '2rem', 0]}>
      <form
        onSubmit={event => {
          event.preventDefault();

          dispatch(
            addTransaction({
              amount: type === 'incoming' ? amount : -Math.abs(amount),
              dueDayNumber: isDistributedDaily ? null : dueDayNumber,
              isDistributedDaily,
              name,
              type,
            }),
          );

          setAmount(DEFAULT_AMOUNT);
          setDueDayNumber(DEFAULT_DUE_DAY_NUMBER);
          setIsDistributedDaily(DEFAULT_IS_DISTRIBUTED_DAILY);
          setName(DEFAULT_NAME);
          setType(DEFAULT_TYPE);
        }}
      >
        <Heading as='h3' display={['none', 'none', 'block']} size='md'>
          Add transaction
        </Heading>
        <FormControl {...formItemProps}>
          <FormLabel>Name:</FormLabel>
          <Input
            isRequired
            // @ts-ignore
            onChange={({ target }) => {
              setName(target.value);
            }}
            placeholder='e.g. "Electricity Bill"'
            type='text'
            value={name}
          />
        </FormControl>
        <FormControl {...formItemProps}>
          <FormLabel>Type:</FormLabel>
          <Select
            onChange={({ target }) => {
              setType(target.value as Transaction['type']);
            }}
            value={type}
          >
            <option value={'incoming'}>Incoming</option>
            <option value={'outgoing'}>Outgoing</option>
          </Select>
        </FormControl>
        <FormControl {...formItemProps}>
          <FormLabel>Amount in {currency.symbol}:</FormLabel>
          <Box display='flex'>
            <NumberInput
              flex={1}
              marginRight='2rem'
              max={Number.MAX_SAFE_INTEGER}
              min={0.01}
              // @ts-ignore
              onChange={(amount: number) => {
                setAmount(amount);
              }}
              value={amount}
            >
              <NumberInputField />
            </NumberInput>
            <FormHelperText whiteSpace='nowrap'>e.g. 119.99</FormHelperText>
          </Box>
        </FormControl>
        <FormControl {...formItemProps}>
          <FormLabel>Due on day:</FormLabel>
          <Select
            isDisabled={isDistributedDaily}
            onChange={({ target }) => {
              const nextDayNumber: NonNullable<Transaction['dueDayNumber']> = +target.value;

              if (isNaN(nextDayNumber)) {
                return;
              }

              setDueDayNumber(nextDayNumber);
            }}
            value={dueDayNumber}
          >
            {range(1, 32).map(dayNumber => (
              <option key={dayNumber} value={dayNumber}>
                {dayNumber}
              </option>
            ))}
          </Select>
        </FormControl>
        <FormControl display='flex' {...formItemProps}>
          <FormLabel>Distribute over whole month:</FormLabel>
          <Checkbox
            isChecked={isDistributedDaily}
            onChange={({ target }) => {
              setIsDistributedDaily(target.checked);
            }}
            variantColor='teal'
          />
        </FormControl>
        <Button type='submit' variant='outline' variantColor='teal' {...formItemProps}>
          Add
        </Button>
      </form>
    </Box>
  );
};
