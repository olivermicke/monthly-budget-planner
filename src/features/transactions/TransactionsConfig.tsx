import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FormControl, FormLabel, Select } from '@chakra-ui/core';

import { changeCurrency, selectCurrency } from './transactionsSlice';

import { CURRENCIES } from './constants';

import { Currency } from './types';

export const TransactionsConfig = () => {
  const currency = useSelector(selectCurrency);
  const dispatch = useDispatch();

  return (
    <FormControl marginTop={['1rem', '1rem', '2rem']}>
      <FormLabel>Currency:</FormLabel>
      <Select
        onChange={({ target }) => {
          dispatch(changeCurrency(target.value as Currency['name']));
        }}
        value={currency.name}
      >
        {Object.values(CURRENCIES).map(({ name }) => (
          <option key={name} value={name}>
            {name}
          </option>
        ))}
      </Select>
    </FormControl>
  );
};
