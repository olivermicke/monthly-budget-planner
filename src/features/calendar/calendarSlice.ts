import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { range } from 'ramda';

import { slice as transactionsSlice } from '../transactions/transactionsSlice';

import { Day as DayType } from './types';
import { RootState } from '../../store';

export const initialState = {
  config: {
    firstDayNumber: 1 as DayType['number'],
  },
  days: range(1, 32).reduce(
    (days, day) => ({
      ...days,
      [day]: {
        balance: 0.0,
        number: day,
      },
    }),
    {},
  ) as { [key: number]: DayType },
};

export const slice = createSlice({
  name: 'calendar',
  initialState,
  reducers: {
    changeFirstDayNumber(state, action: PayloadAction<DayType['number']>) {
      state.config.firstDayNumber = action.payload;
    },
  },
  extraReducers: {
    [transactionsSlice.actions.addTransaction.type]: addTransaction,
  },
});

const calculateSortedDayNumbers = (firstDay: number, dueDay: number | null): number[] => {
  if (firstDay === 1) {
    return range(dueDay ?? 1, 32);
  }

  if (dueDay === null) {
    return [...range(firstDay, 32), ...range(1, firstDay)];
  }

  if (dueDay < firstDay) {
    return range(dueDay, firstDay);
  } else {
    return [...range(dueDay, 32), ...range(1, firstDay)];
  }
};

export function addTransaction(
  state: typeof initialState,
  action: ReturnType<typeof transactionsSlice.actions.addTransaction>,
) {
  const { firstDayNumber } = state.config;
  const { amount, dueDayNumber, isDistributedDaily, type } = action.payload;

  const fixedAmount = type === 'incoming' ? amount : -Math.abs(amount);

  if (isDistributedDaily) {
    const sortedDayNumbers = calculateSortedDayNumbers(firstDayNumber, dueDayNumber);

    const dailyAmount = +(fixedAmount / 31).toFixed(2);

    sortedDayNumbers.forEach((dayNumber, index) => {
      state.days[dayNumber].balance += dailyAmount * (index + 1);
    });
    return;
  }

  if (dueDayNumber) {
    const sortedDayNumbers = calculateSortedDayNumbers(firstDayNumber, dueDayNumber);

    sortedDayNumbers.forEach(dayNumber => {
      state.days[dayNumber].balance += fixedAmount;
    });
    return;
  }
}

export const selectDay = (state: RootState, dayNumber: DayType['number']) => state.calendar.days[dayNumber];
export const selectFirstDayNumber = (state: RootState) => state.calendar.config.firstDayNumber;

export const { changeFirstDayNumber } = slice.actions;

export const reducer = slice.reducer;
