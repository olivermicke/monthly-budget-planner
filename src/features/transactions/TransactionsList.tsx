import React from 'react';
import { useSelector } from 'react-redux';
import { Box, Heading } from '@chakra-ui/core';

import { Transaction } from './Transaction';
import { TransactionsConfig } from './TransactionsConfig';

import { selectTransactions } from './transactionsSlice';

const boxProps = { marginTop: ['1rem', '1rem', '2rem'] };

export const TransactionsList = () => {
  const transactions = useSelector(selectTransactions);

  const transactionsValues = Object.values(transactions);

  return (
    <Box>
      <Heading as='h3' display={['none', 'none', 'block']} size='md'>
        Transactions
      </Heading>

      <TransactionsConfig />

      {transactionsValues.length === 0 ? (
        <Box marginTop='2rem' {...boxProps}>
          No transactions added yet.
        </Box>
      ) : (
        transactionsValues.map(transaction => (
          <Box marginTop='2rem'>
            <Transaction key={transaction.name} {...transaction} />
          </Box>
        ))
      )}
    </Box>
  );
};
