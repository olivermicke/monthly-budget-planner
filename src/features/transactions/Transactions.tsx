import React, { FunctionComponent } from 'react';
import {
  Accordion,
  AccordionHeader,
  AccordionIcon,
  AccordionItem,
  AccordionPanel,
  Box,
  Divider,
} from '@chakra-ui/core';

import { TransactionForm } from './TransactionForm';
import { TransactionsList } from './TransactionsList';

export const Transactions: FunctionComponent<{}> = () => (
  <>
    <Box display={['block', 'block', 'none']}>
      <Accordion allowMultiple allowToggle>
        <AccordionItem>
          <AccordionHeader>
            <Box flex='1' textAlign='left'>
              Add transaction
            </Box>
            <AccordionIcon />
          </AccordionHeader>
          <AccordionPanel>
            <TransactionForm />
          </AccordionPanel>
        </AccordionItem>

        <AccordionItem>
          <AccordionHeader>
            <Box flex='1' textAlign='left'>
              List of transactions
            </Box>
            <AccordionIcon />
          </AccordionHeader>
          <AccordionPanel>
            <TransactionsList />
          </AccordionPanel>
        </AccordionItem>
      </Accordion>
    </Box>

    <Box display={['none', 'none', 'block']}>
      <Box display='flex' marginBottom='2rem'>
        <Box width='50%' paddingRight='6rem'>
          <TransactionForm />
        </Box>
        <Divider orientation='vertical' />
        <Box width='50%' paddingLeft='6rem'>
          <TransactionsList />
        </Box>
      </Box>
    </Box>
  </>
);
