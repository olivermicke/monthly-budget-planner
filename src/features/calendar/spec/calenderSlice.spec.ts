import { changeFirstDayNumber, initialState, reducer, selectFirstDayNumber } from '../calendarSlice';
import { slice as transactionsSlice } from '../../transactions/transactionsSlice';

import { RootState } from '../../../store';

describe('calendarSlice', () => {
  describe('addTransaction', () => {
    describe('handles transactions distributed over the month', () => {
      test('when first day is 1', () => {
        const payload = {
          amount: 310,
          dueDayNumber: null,
          isDistributedDaily: true,
          name: 'foo',
          type: 'incoming',
        } as const;

        const action = {
          payload,
          type: transactionsSlice.actions.addTransaction.type,
        };

        const nextState = reducer(initialState, action);
        const balances = Object.values(nextState.days).map(day => day.balance);

        expect(balances).toMatchSnapshot();
      });

      test('when first day is not 1', () => {
        const modifiedInitialState = {
          ...initialState,
          config: {
            ...initialState.config,
            firstDayNumber: 5,
          },
        };

        const payload = {
          amount: 310,
          dueDayNumber: null,
          isDistributedDaily: true,
          name: 'foo',
          type: 'incoming',
        } as const;

        const action = {
          payload,
          type: transactionsSlice.actions.addTransaction.type,
        };

        const nextState = reducer(modifiedInitialState, action);
        const balances = Object.values(nextState.days).map(day => day.balance);

        expect(balances).toMatchSnapshot();
      });
    });

    describe('handles transactions on a given day', () => {
      test('when first day is 1', () => {
        const payload = {
          amount: 10,
          dueDayNumber: 2,
          isDistributedDaily: false,
          name: 'foo',
          type: 'incoming',
        } as const;

        const action = {
          payload,
          type: transactionsSlice.actions.addTransaction.type,
        };

        const nextState = reducer(initialState, action);
        const balances = Object.values(nextState.days).map(day => day.balance);

        const expectedBalances = [0, ...new Array(30).fill(payload.amount)];

        expect(balances).toEqual(expectedBalances);
      });

      test('when first day is not 1 and firstDay > dueDay', () => {
        const modifiedInitialState = {
          ...initialState,
          config: {
            ...initialState.config,
            firstDayNumber: 5,
          },
        };

        const payload = {
          amount: 10,
          dueDayNumber: 2,
          isDistributedDaily: false,
          name: 'foo',
          type: 'incoming',
        } as const;

        const action = {
          payload,
          type: transactionsSlice.actions.addTransaction.type,
        };

        const nextState = reducer(modifiedInitialState, action);
        const balances = Object.values(nextState.days).map(day => day.balance);

        const expectedBalances = [0, ...new Array(3).fill(payload.amount), ...new Array(27).fill(0)];

        expect(balances).toEqual(expectedBalances);
      });

      test('when first day is not 1 and firstDay < dueDay', () => {
        const modifiedInitialState = {
          ...initialState,
          config: {
            ...initialState.config,
            firstDayNumber: 2,
          },
        };

        const payload = {
          amount: 10,
          dueDayNumber: 5,
          isDistributedDaily: false,
          name: 'foo',
          type: 'incoming',
        } as const;

        const action = {
          payload,
          type: transactionsSlice.actions.addTransaction.type,
        };

        const nextState = reducer(modifiedInitialState, action);
        const balances = Object.values(nextState.days).map(day => day.balance);

        const expectedBalances = [payload.amount, ...new Array(3).fill(0), ...new Array(27).fill(payload.amount)];

        expect(balances).toEqual(expectedBalances);
      });
    });
  });

  test('changeFirstDayNumber', () => {
    const nextState = reducer(initialState, changeFirstDayNumber(5));
    const rootState = { calendar: nextState } as RootState;
    expect(selectFirstDayNumber(rootState)).toEqual(5);
  });
});
