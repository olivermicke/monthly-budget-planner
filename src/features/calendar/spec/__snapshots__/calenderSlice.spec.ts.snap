// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendarSlice addTransaction handles transactions distributed over the month when first day is 1 1`] = `
Array [
  10,
  20,
  30,
  40,
  50,
  60,
  70,
  80,
  90,
  100,
  110,
  120,
  130,
  140,
  150,
  160,
  170,
  180,
  190,
  200,
  210,
  220,
  230,
  240,
  250,
  260,
  270,
  280,
  290,
  300,
  310,
]
`;

exports[`calendarSlice addTransaction handles transactions distributed over the month when first day is not 1 1`] = `
Array [
  280,
  290,
  300,
  310,
  10,
  20,
  30,
  40,
  50,
  60,
  70,
  80,
  90,
  100,
  110,
  120,
  130,
  140,
  150,
  160,
  170,
  180,
  190,
  200,
  210,
  220,
  230,
  240,
  250,
  260,
  270,
]
`;
